File: ann.cpp
#include<iostream>
#include<vector>
#include<cmath>
#include<random>
#include <fstream>
#include <sstream>
#include <string>
#include <stdexcept>

using namespace std;

class Matrix {
public:
    vector<vector<double>> data;
    int rows, cols;

    Matrix(int rows, int cols);
    Matrix() : rows(0), cols(0) {};

    // Hàm khởi tạo ngẫu nhiên
    void randomize();
    void initialBias();


    // Các phép toán ma trận
    Matrix transpose();
    Matrix add(const Matrix& other) const;
    Matrix subtract(const Matrix& other) const;
    Matrix multiply(const Matrix& other) const;
    Matrix multiplyGrad(const Matrix& other,int m) const;
    Matrix scalarMultiply(double scalar) const;
    Matrix updateBias(int m) const;
    int argmax(int row) const;

    // Hàm kích hoạt ReLU
    void applyReLU();

    // Đạo hàm hàm kích hoạt ReLU
    void applyReLUDerivative();

    // Hàm Softmax
    void applySoftmax();

    // In ma trận
    void print() const;
};

Matrix::Matrix(int rows, int cols) {
    this->rows = rows;
    this->cols = cols;
    data.resize(rows, vector<double>(cols, 0));
}



// Hàm khởi tạo ngẫu nhiên
void Matrix::randomize() {
    random_device rd;
    mt19937 gen(rd());

    // He initialization
    double scale = sqrt(2.0 / rows);
    normal_distribution<double> d(0, scale);

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            data[i][j] = d(gen);
        }
    }
}

// Hàm khởi tạo ngẫu nhiên
void Matrix::initialBias() {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            data[i][j] = 0.01;
        }
    }
}

Matrix Matrix::transpose() {
    Matrix result(cols, rows);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            result.data[j][i] = data[i][j];
        }
    }
    return result;
}

// Hàm cộng ma trận
Matrix Matrix::add(const Matrix& other) const {
    if (rows != other.rows && other.rows != 1) {
        throw invalid_argument("Matrix dimensions must match or the other matrix must have one row for broadcasting.");
    }
    Matrix result(rows, cols);

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            result.data[i][j] = data[i][j] + other.data[0][j];
        }
    }
    return result;
}

// Hàm trừ ma trận
Matrix Matrix::subtract(const Matrix& other) const {
    if (rows != other.rows || cols != other.cols) {
        throw invalid_argument("Matrix dimensions must match for subtraction.");
    }
    Matrix result(rows, cols);
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            result.data[i][j] = data[i][j] - other.data[i][j];
        }
    }
    return result;
}

// Hàm nhân ma trận
Matrix Matrix::multiply(const Matrix& other) const {
    if (cols != other.rows) {
        throw invalid_argument("Matrix dimensions do not match for multiplication.");
    }

    Matrix result(rows, other.cols);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < other.cols; j++) {
            result.data[i][j] = 0; // Initialize the result cell
            for (int k = 0; k < cols; k++) {
                result.data[i][j] += data[i][k] * other.data[k][j];
            }
        }
    }
    return result;
}

// Hàm nhân ma trận để tính gradient
Matrix Matrix::multiplyGrad(const Matrix& other, int m ) const {
    if (cols != other.rows) {
        throw invalid_argument("Matrix dimensions do not match for multiplication.");
    }

    Matrix result(rows, other.cols);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < other.cols; j++) {
            result.data[i][j] = 0; // Initialize the result cell
            for (int k = 0; k < cols; k++) {
                result.data[i][j] += data[i][k] * other.data[k][j];
            }
            result.data[i][j] /= m;
        }
    }
    return result;
}

// Hàm Update bias
Matrix Matrix::updateBias(int m) const {
    Matrix result(1, cols);

    for (int j = 0; j < cols; j++) {
        double sumCols = 0.0;
        for (int i = 0; i < rows; i++) {
            sumCols += data[i][j]; // Cộng các phần tử trong cột j
        }
        // Gán giá trị tổng chia cho số hàng (rows) vào ma trận kết quả
        result.data[0][j] = sumCols / m;
    }
    return result;
}


// Hàm nhân với số vô hướng
Matrix Matrix::scalarMultiply(double scalar) const {
    Matrix result(rows, cols);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            result.data[i][j] = data[i][j] * scalar;
        }
    }
    return result;
}

// Hàm ReLU
void Matrix::applyReLU() {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            data[i][j] = max(0.0, data[i][j]);
        }
    }
}

// Đạo hàm ReLU
void Matrix::applyReLUDerivative() {
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            // Áp dụng đạo hàm ReLU: nếu giá trị > 0, đạo hàm = 1, nếu <= 0, đạo hàm = 0
            data[i][j] = data[i][j] > 0 ? 1.0 : 0.0;
        }
    }
}

// Hàm Softmax
void Matrix::applySoftmax() {
    for (int i = 0; i < rows; i++) {
        double max_val = data[i][0];
        for (int j = 1; j < cols; j++) {
            max_val = max(max_val, data[i][j]);
        }

        double sum_exp = 0.0;
        for (int j = 0; j < cols; j++) {
            sum_exp += exp(data[i][j] - max_val);
        }

        for (int j = 0; j < cols; j++) {
            data[i][j] = exp(data[i][j] - max_val) / sum_exp;
        }
    }
}
// In ma trận
void Matrix::print() const {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cout << data[i][j] << " ";
        }
        cout << endl;
    }
}

int Matrix::argmax(int row) const {
    if (row < 0 || row >= rows) {
        throw std::out_of_range("Row index out of range");
    }

    int max_index = 0;
    double max_value = data[row][0];

    for (int col = 1; col < cols; col++) {
        if (data[row][col] > max_value) {
            max_value = data[row][col];
            max_index = col;
        }
    }
    return max_index;
}


// Tính Cross-Entropy Loss
double crossEntropyLoss(const Matrix& predictions, const Matrix& targets) {
    double loss = 0.0;
    for (int i = 0; i < targets.rows; i++) {
        // Find the index of the maximum value in the target (one-hot encoded)
        int targetIndex = 0;
        for (int j = 1; j < targets.cols; j++) {
            if (targets.data[i][j] > targets.data[i][targetIndex]) {
                targetIndex = j;
            }
        }

        // Compute the log likelihood
        double logLikelihood = -log(predictions.data[i][targetIndex]);
        loss += logLikelihood;
    }

    // Return the average loss over all samples
    return loss / targets.rows;
}

// Modified average loss calculation in train method
double calculateAverageLoss(const Matrix& lossMatrix) {
    double totalLoss = 0.0;
    int count = 0;

    for (int i = 0; i < lossMatrix.rows; ++i) {
        for (int j = 0; j < lossMatrix.cols; ++j) {
            if (lossMatrix.data[i][j] != 0) { // Only count non-zero losses
                totalLoss -= lossMatrix.data[i][j];
                count++;
            }
        }
    }

    return count > 0 ? totalLoss / count : 0.0;
}


class NeuralNetwork {
public:
    Matrix W1, b1, W_hidden2, b_hidden2, W2, b2;
    int batch_size;

    NeuralNetwork(int input_size, int hidden_size1, int hidden_size2, int output_size, int batch_size)
        : W1(input_size, hidden_size1), b1(1, hidden_size1),
        W_hidden2(hidden_size1, hidden_size2), b_hidden2(1, hidden_size2),
        W2(hidden_size2, output_size), b2(1, output_size) {

        W1.randomize();
        //b1.initialBias();
        

        W_hidden2.randomize();
        //b_hidden2.initialBias();

        W2.randomize();


   
    }

    void train(const Matrix& input, const Matrix& target, double learning_rate, int epochs);
};

void NeuralNetwork::train(const Matrix& input, const Matrix& target, double learning_rate, int epochs) {
    Matrix X = input;  // Copy dữ liệu vào một biến mới

    int m = X.rows;

    for (int epoch = 0; epoch < epochs; epoch++) {
        // Forward pass
        // Lớp ẩn đầu tiên
        Matrix Z1 = X.multiply(W1).add(b1);
        Z1.applyReLU();

        // Lớp ẩn thứ hai
        Matrix Z2_hidden = Z1.multiply(W_hidden2).add(b_hidden2);
        Z2_hidden.applyReLU();

        // Lớp đầu ra
        Matrix Z2 = Z2_hidden.multiply(W2).add(b2);
        Z2.applySoftmax();

        // Tính mất mát
        double loss = crossEntropyLoss(Z2, target);
      
        // Tính độ chính xác
        int correct_predictions = 0;
        for (int i = 0; i < Z2.rows; i++) {

            if (Z2.argmax(i) == target.argmax(i)) {
                correct_predictions++;
            }
        }
        double accuracy = static_cast<double>(correct_predictions) / Z2.rows;

        cout << "Epoch " << epoch << ", Loss: " << loss << ", Accuracy: " << accuracy * 100 << "%" << endl;
   

       // Backpropagation
        Matrix dZ_output = Z2.subtract(target); // Gradient tại output layer
        Matrix dW_output = Z2_hidden.transpose().multiplyGrad(dZ_output,m);
        Matrix db_output = dZ_output.updateBias(m);

        // Gradient tại hidden layer 2
        Matrix dZ_hidden2 = dZ_output.multiplyGrad(W2.transpose(),m);
        dZ_hidden2.applyReLUDerivative();

        Matrix dW_hidden2 = Z1.transpose().multiplyGrad(dZ_hidden2,m);
        Matrix db_hidden2 = dZ_hidden2.updateBias(m);

        // Gradient tại hidden layer 1
        Matrix dZ_hidden1 = dZ_hidden2.multiplyGrad(W_hidden2.transpose(),m);
        dZ_hidden1.applyReLUDerivative();

        Matrix dW_hidden1 = X.transpose().multiplyGrad(dZ_hidden1,m);
        Matrix db_hidden1 = dZ_hidden1.updateBias(m);

        // Cập nhật trọng số và bias
        W1 = W1.subtract(dW_hidden1.scalarMultiply(learning_rate));
        b1 = b1.subtract(db_hidden1.scalarMultiply(learning_rate));
        W_hidden2 = W_hidden2.subtract(dW_hidden2.scalarMultiply(learning_rate));
        b_hidden2 = b_hidden2.subtract(db_hidden2.scalarMultiply(learning_rate));
        W2 = W2.subtract(dW_output.scalarMultiply(learning_rate));
        b2 = b2.subtract(db_output.scalarMultiply(learning_rate));

    }
}



void loadData(const string& filename, Matrix& X, Matrix& Y, int sample) {
    try {
        ifstream file(filename);
        if (!file.is_open()) {
            throw runtime_error("Error: Could not open file " + filename);
        }

        string line;

        // Skip the header line (first line)
        if (!getline(file, line)) {
            throw runtime_error("Error: The file is empty or cannot read the header.");
        }

        int row = 0;

        // Process the rest of the file
        while (getline(file, line) && (row < sample)) {

            stringstream ss(line);
            string value;
            vector<double> image_pixels;
            string label;

            // Đọc nhãn (label) của ảnh
            if (!getline(ss, label, ',')) {
                throw runtime_error("Error: Invalid label format in line " + to_string(row + 2)); // +2 to account for the header
            }

            // Đọc các giá trị pixel của ảnh
            while (getline(ss, value, ',')) {
                try {
                    image_pixels.push_back(stod(value));  // Chuyển string thành double
                }
                catch (const invalid_argument& e) {
                    throw runtime_error("Error: Invalid pixel value in line " + to_string(row + 2)); // +2 to account for the header
                }
            }

            // Lưu nhãn vào ma trận Y
            for (int i = 0; i < Y.cols; ++i) {
                Y.data[row][i] = (i == stoi(label)) ? 1.0 : 0.0;
            }

            // Lưu các pixel vào ma trận X
            if (image_pixels.size() != X.cols) {
                throw runtime_error("Error: Mismatched number of pixels in line " + to_string(row + 2)); // +2 to account for the header
            }

            for (int i = 0; i < X.cols; ++i) {
                
                X.data[row][i] = image_pixels[i] / 255.0;
            }

            ++row;
        }

        if (row == 0) {
            throw runtime_error("Error: No data found in file " + filename);
        }

        file.close();
    }
    catch (const exception& e) {
        cerr << e.what() << endl;
    }
}

void saveModel(const NeuralNetwork& nn, const string& filename) {
    ofstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Could not open file to save model." << endl;
        return;
    }

    auto saveMatrix = [&file](const Matrix& mat) {
        file << mat.rows << " " << mat.cols << endl;
        for (const auto& row : mat.data) {
            for (double val : row) {
                file << val << " ";
            }
            file << endl;
        }
        };

    saveMatrix(nn.W1);
    saveMatrix(nn.b1);
    saveMatrix(nn.W_hidden2);
    saveMatrix(nn.b_hidden2);
    saveMatrix(nn.W2);
    saveMatrix(nn.b2);

    file.close();
    cout << "Model saved to " << filename << endl;
}

void loadModel(NeuralNetwork& nn, const string& filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Could not open file to load model." << endl;
        return;
    }

    auto loadMatrix = [&file](Matrix& mat) {
        file >> mat.rows >> mat.cols;
        mat.data.resize(mat.rows, vector<double>(mat.cols));
        for (auto& row : mat.data) {
            for (double& val : row) {
                file >> val;
            }
        }
        };

    loadMatrix(nn.W1);
    loadMatrix(nn.b1);
    loadMatrix(nn.W_hidden2);
    loadMatrix(nn.b_hidden2);
    loadMatrix(nn.W2);
    loadMatrix(nn.b2);

    file.close();
    cout << "Model loaded from " << filename << endl;
}

void predict(const NeuralNetwork& nn, const Matrix& input, Matrix& output) {
    Matrix X = input;

    // Forward pass
    Matrix Z1 = X.multiply(nn.W1).add(nn.b1);
    Z1.applyReLU();

    Matrix Z2_hidden = Z1.multiply(nn.W_hidden2).add(nn.b_hidden2);
    Z2_hidden.applyReLU();

    Matrix Z2 = Z2_hidden.multiply(nn.W2).add(nn.b2);
    Z2.applySoftmax();

    output = Z2;
}

void evaluate(const NeuralNetwork& nn, const Matrix& X_test, const Matrix& Y_test) {
    Matrix predictions;
    predict(nn, X_test, predictions);

    int correct_predictions = 0;
    for (int i = 0; i < predictions.rows; ++i) {
        if (predictions.argmax(i) == Y_test.argmax(i)) {
            correct_predictions++;
        }
    }
    double accuracy = static_cast<double>(correct_predictions) / predictions.rows;
    cout << "Accuracy on test set: " << accuracy * 100 << "%" << endl;
}


int main() {
    int input_size = 784;  // 28x28 pixel cho mỗi ảnh
    int hidden_size1 = 128;
    int hidden_size2 = 128;
    int output_size = 10;  // 10 lớp cho 10 nhãn
    int batch_size = 200;

    Matrix X_train(batch_size, input_size);  // Tập huấn luyện (60000 ảnh)
    Matrix Y_train(batch_size, output_size); // Nhãn tương ứng

    // Đọc dữ liệu từ file CSV
    loadData("fashion-mnist_train.csv", X_train, Y_train, batch_size);

    cout << "X_train shape: " << X_train.rows << " " << X_train.cols << endl;
    cout << "Y_train shape: " << Y_train.rows << " " << Y_train.cols << endl;
    // Tạo và huấn luyện mạng
    
    NeuralNetwork nn(input_size, hidden_size1,hidden_size2, output_size, batch_size);
    nn.train(X_train, Y_train, 0.01, 100);
    
    //saveModel(nn, "saved_model.txt");

    batch_size = 200;
    // Testing model
    Matrix X_test(batch_size, input_size);  // 10000 ảnh kiểm tra
    Matrix Y_test(batch_size, output_size);

    loadData("fashion-mnist_test.csv", X_test, Y_test, batch_size);

    // Tải mô hình và đánh giá
    //NeuralNetwork nnTest(input_size, hidden_size1, hidden_size2, output_size, batch_size);
    //loadModel(nnTest, "saved_model.txt");

    evaluate(nn, X_test, Y_test);

    return 0;
}

File: ann.py
import numpy as np
# Import fashion_mnist
from tensorflow.keras.datasets import fashion_mnist
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

def load_data():
    # Load Fashion MNIST directly
    (X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()
    
    # Normalize the data
    X_train = X_train.reshape(-1, 28 * 28).astype(np.float32) / 255.0
    X_test = X_test.reshape(-1, 28 * 28).astype(np.float32) / 255.0
    return X_train, y_train, X_test, y_test

# One-hot encoding for labels
def one_hot_encode(y, num_classes):
    encoder = OneHotEncoder(sparse_output=False, categories='auto')
    y = y.reshape(-1, 1)
    return encoder.fit_transform(y)

# Accuracy calculation
def calculate_accuracy(predictions, true_labels):
    return np.mean(predictions == true_labels) * 100

# Split data into train and validation
def split_data(X, y, validation_size=0.2):
    return train_test_split(X, y, test_size=validation_size, random_state=42)

# Activation Functions
def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

def softmax(x):
    exp_x = np.exp(x - np.max(x, axis=1, keepdims=True))
    return exp_x / np.sum(exp_x, axis=1, keepdims=True)

# Loss Function
def cross_entropy_loss(predictions, targets):
    n_samples = targets.shape[0]
    log_likelihood = -np.log(predictions[range(n_samples), targets.argmax(axis=1)])
    return np.sum(log_likelihood) / n_samples

# Gradient of Loss Function
def gradient_cross_entropy(predictions, targets):
    return predictions - targets

# Initialize weights and biases using He Initialization
def initialize_weights(input_size, hidden1_size, hidden2_size, output_size):
    weights = {
        'W1': np.random.randn(input_size, hidden1_size) * np.sqrt(2. / input_size),
        'b1': np.zeros((1, hidden1_size)),
        'W2': np.random.randn(hidden1_size, hidden2_size) * np.sqrt(2. / hidden1_size),
        'b2': np.zeros((1, hidden2_size)),
        'W3': np.random.randn(hidden2_size, output_size) * np.sqrt(2. / hidden2_size),
        'b3': np.zeros((1, output_size))
    }
    return weights


# Forward Propagation
def forward_propagation(X, weights):
    Z1 = np.dot(X, weights['W1']) + weights['b1']
    A1 = relu(Z1)
    Z2 = np.dot(A1, weights['W2']) + weights['b2']
    A2 = relu(Z2)
    Z3 = np.dot(A2, weights['W3']) + weights['b3']
    A3 = softmax(Z3)
    cache = {'Z1': Z1, 'A1': A1, 'Z2': Z2, 'A2': A2, 'Z3': Z3, 'A3': A3}
    return A3, cache

# Backward Propagation
def backward_propagation(X, Y, cache, weights,loss_entropy):
    m = X.shape[0]
    gradients = {}
    
    dZ3 = gradient_cross_entropy(cache['A3'], Y)* loss_entropy
    gradients['dW3'] = np.dot(cache['A2'].T, dZ3) / m
    gradients['db3'] = np.sum(dZ3, axis=0, keepdims=True) / m

    
    dA2 = np.dot(dZ3, weights['W3'].T)
    dZ2 = dA2 * relu_derivative(cache['Z2'])
    gradients['dW2'] = np.dot(cache['A1'].T, dZ2) / m
    gradients['db2'] = np.sum(dZ2, axis=0, keepdims=True) / m
    
    dA1 = np.dot(dZ2, weights['W2'].T)
    dZ1 = dA1 * relu_derivative(cache['Z1'])
    gradients['dW1'] = np.dot(X.T, dZ1) / m
    gradients['db1'] = np.sum(dZ1, axis=0, keepdims=True) / m
    
    return gradients

# Update Weights
def update_weights(weights, gradients, learning_rate):
    for key in weights.keys():
        weights[key] -= learning_rate * gradients['d' + key]
    return weights

# Training Loop (with accuracy)
def train(X_train, y_train, X_val, y_val, epochs, learning_rate, input_size, hidden1_size, hidden2_size, output_size):
    weights = initialize_weights(input_size, hidden1_size, hidden2_size, output_size)
    
    for epoch in range(epochs):
        # Forward propagation
        predictions, cache = forward_propagation(X_train, weights)
        
        # Compute training loss
        train_loss = cross_entropy_loss(predictions, y_train)
        train_accuracy = calculate_accuracy(np.argmax(predictions, axis=1), np.argmax(y_train, axis=1))
        
        # Backward propagation
        gradients = backward_propagation(X_train, y_train, cache, weights,train_loss)
        
        # Update weights
        weights = update_weights(weights, gradients, learning_rate)
        
        # Validation
        val_predictions, _ = forward_propagation(X_val, weights)
        val_loss = cross_entropy_loss(val_predictions, y_val)
        val_accuracy = calculate_accuracy(np.argmax(val_predictions, axis=1), np.argmax(y_val, axis=1))
        
        if epoch % 10 == 0:
            print(f"Epoch {epoch}/{epochs}, Train Loss: {train_loss:.4f}, Train Acc: {train_accuracy:.2f}%, "
                  f"Val Loss: {val_loss:.4f}, Val Acc: {val_accuracy:.2f}%")
    
    return weights


# Predict function
def predict(X, weights):
    predictions, _ = forward_propagation(X, weights)
    return np.argmax(predictions, axis=1)

# Accuracy calculation
def calculate_accuracy(predictions, true_labels):
    return np.mean(predictions == true_labels) * 100
# Main Function
if __name__ == "__main__":
    # Load data
    X_train, y_train, X_test, y_test = load_data()
    
    # One-hot encode the labels
    y_train_one_hot = one_hot_encode(y_train, num_classes=10)
    y_test_one_hot = one_hot_encode(y_test, num_classes=10)
    
    print("Training data shape:", X_train.shape, y_train_one_hot.shape)
    print("Test data shape:", X_test.shape, y_test_one_hot.shape)
    
    # Split into train and validation sets
    X_train_split, X_val, y_train_split, y_val = split_data(X_train, y_train_one_hot)
    
    print("Train split shape:", X_train_split.shape, y_train_split.shape)
    print("Validation split shape:", X_val.shape, y_val.shape)
    
    # Hyperparameters
    input_size = 28 * 28
    hidden1_size = 128
    hidden2_size = 128
    output_size = 10
    epochs = 50
    learning_rate = 0.1
    
    # Train model
    trained_weights = train(X_train_split, y_train_split, X_val, y_val, epochs, learning_rate, input_size, hidden1_size, hidden2_size, output_size)
    
    # Test model accuracy
    test_predictions = predict(X_test, trained_weights)
    test_accuracy = calculate_accuracy(test_predictions, y_test)
    print(f"Test Accuracy: {test_accuracy:.2f}%")
